internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+package rpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"connectrpc.com/connect"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+const (
+	otelFtlRequestKey       = "ftl.requestKey"
+	otelFtlVerbRef          = "ftl.verb.ref"
+	otelFtlVerbModule       = "ftl.verb.module"
+	otelMessageEvent        = "message"
+	otelMessageIDKey        = attribute.Key("message.id")
+	otelMessageSizeKey      = attribute.Key("message.uncompressed_size")
+	otelMessageTypeKey      = attribute.Key("message.type")
+	otelMessageTypeSent     = "SENT"
+	otelMessageTypeReceived = "RECEIVED"
+)
+
+func OtelInterceptor() connect.Interceptor {
+	return &otelInterceptor{}
+}
+
+type otelInterceptor struct{}
+
+var clientInstruments instrumentation
+var serverInstruments instrumentation
+
+func init() {
+	clientInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.client"))
+	serverInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.server"))
+}
+
+func getInstruments(isClient bool) instrumentation {
+	if isClient {
+		return clientInstruments
+	}
+	return serverInstruments
+}
+
+func getAttributes(ctx context.Context, rpcSystemKey string) []attribute.KeyValue {
+	logger := log.FromContext(ctx)
+	attributes := []attribute.KeyValue{
+		semconv.RPCSystemKey.String(rpcSystemKey),
+	}
+	requestKey, err := RequestKeyFromContext(ctx)
+	if err != nil {
+		logger.Debugf("failed to get request key: %s", err)
```

Comment:
Since request key is so valuable, do you think we should raise this log level to Warn or even Error? I'm not sure personally

Reply:
The err case is only really hit if there is a request key that we're unable to parse. It is indicative of an issue though, so good call bumping the sev